pennylane/devices/task_qubit.py

```diff
@@ -0,0 +1,243 @@
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains a proxy qubit object for spawning multiple instances of 
+a given qubit type for run on a background scheduler.
+"""
+from typing import List, Union, Tuple, Dict
+from contextlib import nullcontext
+import pennylane as qml
+from pennylane import QubitDevice
+from .default_qubit import DefaultQubit
+from .._version import __version__
+
+try:
+    import dask
+    import dask.distributed as dist
+    from dask.distributed import worker_client, performance_report
+
+    from distributed.protocol import (
+        dask_serialize,
+        dask_deserialize,
+    )
+
+    @dask_serialize.register(qml.numpy.tensor)
+    def serialize(tensor: qml.numpy.tensor) -> Tuple[Dict, List[bytes]]:
+        "Defines a serializer for the Dask backend"
+        header, frames = dist.protocol.numpy.serialize_numpy_ndarray(tensor)
+        header["type"] = "qml.numpy.tensor"
+        header["requires_grad"] = tensor.requires_grad
+        frames = [tensor.data]
+        return header, frames
+
+    @dask_deserialize.register(qml.numpy.tensor)
+    def deserialize(header: Dict, frames: List[bytes]) -> qml.numpy.tensor:
+        "Defines a deserializer for the Dask backend"
+        return qml.numpy.tensor(frames[0], requires_grad=header["requires_grad"])
+
+
+except ImportError as e:  # pragma: no cover
+    raise ImportError("task.qubit requires installing dask and dask.distributed") from e
+
+
+class ProxyInstanceCLS(classmethod):
+    """
+    This utility class allows the use of both an instance
+    as well as class method types. For situations where
+    the explicit class method is needed, we can use this as a
+    decorator, and select the appropriate call dynamically.
+    This is an essential support for accessing backend
+    functionality supports with the proxy `task.qubit`.
+
+    The implementation is based on the Python descriptor guide
+    as mentioned https://stackoverflow.com/questions/28237955/same-name-for-classmethod-and-instancemethod/28238047#28238047
+    """
+
+    def __get__(self, instance, type_):
+        descr_get = super().__get__ if instance is None else self.__func__.__get__
+        return descr_get(instance, type_)
+
+
+# pylint: too-few-public-methods, bad-classmethod-argument
+class TaskQubit(QubitDevice):
+    """Proxy simulator plugin written using Dask.Distributed as a task-distribution scheduling backend.
+
+    **Short name:** ``task.qubit``
+
+    This device provides a pure-state qubit simulator wrapping both ``"default.qubit"`` and ``"lightning.qubit"``,
+    and written to allow batched offloading to a Dask scheduler.
```

Comment:
Huh. This almost makes me wonder if something like


```python
>>> dev = qml.device("lightning.qubit")
>>> task_device = qml.taskify(dev)
```

is possible 🤔 🤔 🤔

Reply:
Oh, that seems like a great idea. Even if it doesn't explicitly work now, getting it into this state would be quite useful.

Reply:
What needs to be done to support this UI? From reading the code below, it seems that only two things:

1. The ability to pass new configuration options. This should be supported via

   ```python
   >>> task_device = qml.taskify(dev, some_task_option=True)
   ```

2. We need to copy `dev`, and in the copy override the `batch_execute` method. This is easy to do if we simply have

   ```python
   >>> dev_new = copy.copy(dev)
   >>> dev_new.batch_execute = types.MethodType(task_batch_execute, dev_new)
   ```

   although I always feel a bit dirty using `copy` (although I guess nothing _wrong_ with it right?). Maybe there is also a metaclassy-approach to doing this.

3. We need to be able to make new instances of a device given an instantiated device. Also should be pretty easy, since we can just call `dev.__class__`

Reply:
The benefits here is that you don't need to worry about subclassing `Device` any more, the issues re: the capabilities dictionary disappear, etc.

Reply:
instead you just have a function that takes a device and returns a new, task-enabled device.

Reply:
Rather than making copies, would it be simpler to use the `dev` information to instantiate a `TaskQubit` device --- since no memory is needed on the host, this should be cheap, as we should not need to do any copies (assuming you don't need the existing statevector data on every task, in which case this becomes a different issue). Rather than a 1-to-many copy I think a broadcasting of the operations would be cheaper at large-scales.

I have taken a simple pass at this now, just wrapping the call to `TaskQubit` with the device info, but I do agree that moving away from the subclassing to be a better longer-term solution.